Tools/Scripts/libraries/webkitflaskpy/webkitflaskpy/tests/ipc_manager_unittest.py

```diff
@@ -0,0 +1,123 @@
+# Copyright (C) 2023 Apple Inc. All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions
+# are met:
+# 1.  Redistributions of source code must retain the above copyright
+#     notice, this list of conditions and the following disclaimer.
+# 2.  Redistributions in binary form must reproduce the above copyright
+#     notice, this list of conditions and the following disclaimer in the
+#     documentation and/or other materials provided with the distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS "AS IS" AND
+# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS BE LIABLE FOR
+# ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import unittest
+
+from webkitflaskpy import IPCManager
+
+
+class IPCManagerUnittest(unittest.TestCase):
+    COUNTER = 0
+    INCREMENT = 'increment'
+    DECREMENT = 'decrement'
+
+    @classmethod
+    def increment(cls, value=1):
+        if value <= 0:
+            raise ValueError('Increment only accepts positive values')
+        cls.COUNTER += value
+
+    @classmethod
+    def decrement(cls, value=1):
+        if value <= 0:
+            raise ValueError('Decrement only accepts positive values')
+        if value > cls.COUNTER:
+            raise ValueError('Trying to decrement more than available in counter')
+        cls.COUNTER -= value
+
+    def test_enqueue(self):
+        IPCManagerUnittest.COUNTER = 0
+        manager = IPCManager()
+        manager.register(self.INCREMENT, self.increment)
+        manager.register(self.DECREMENT, self.decrement)
+
+        manager.enqueue(self.INCREMENT, 1)
+        self.assertEqual(
+            [('9ffcb5a7b14e4536e69f6527b1283237', ['increment', [1], {}])],
+            list(manager.tasks()),
+        )
+        manager.enqueue(self.INCREMENT, 2)
+        self.assertEqual(2, len(list(manager.tasks())))
+
+        manager.enqueue(self.DECREMENT, 1)
```

Comment:
If I do one more `manager.enqueue(self.INCREMENT, 1)` would the queue size stay unchanged since we enqueued same type of request already?

Reply:
Correct, will add that in this test too.

Reply:
That's an interesting design. Meaning that counter can be different based on the execution order.
```
manager.enqueue(self.INCREMENT, 1)
manager.enqueue(self.INCREMENT, 1)
manager.dequeue('9ffcb5a7b14e4536e69f6527b1283237')
manager.dequeue('9ffcb5a7b14e4536e69f6527b1283237')
```
We will ends with COUNTER=1

If we do
```
manager.enqueue(self.INCREMENT, 1)
manager.dequeue('9ffcb5a7b14e4536e69f6527b1283237')
manager.enqueue(self.INCREMENT, 1)
manager.dequeue('9ffcb5a7b14e4536e69f6527b1283237')
```
We will ends with COUNTER=2

Is our particular use case fine with this behavior?

Reply:
I'm going to enshrine this example in the unit tests.

The short answer to this is "at the moment, and in the immediate future, yes".

The service that uses this now is bug-bridge.webkit.org, and this behavior is acceptable there. Re-doing a task costs compute and network, but is otherwise safe to repeat. If given the option, we'd rather save compute and network.

I plan to use this in commits.webkit.org, where repeating tasks basically means repeating pushes we're already performed, and results.webkit.org, which is designed to be able to re-process uploads.